libraries/shared/storage/backfiller.go

```diff
@@ -0,0 +1,165 @@
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package storage
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/statediff"
+	"github.com/sirupsen/logrus"
+
+	"github.com/vulcanize/vulcanizedb/libraries/shared/fetcher"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/storage/utils"
+)
+
+const (
+	DefaultMaxBatchSize   uint64 = 1000
+	defaultMaxBatchNumber int64  = 10
+)
+
+// BackFiller is the backfilling interface
+type BackFiller interface {
+	BackFill(endingBlock uint64, backFill chan utils.StorageDiff, errChan chan error, done chan bool) error
+}
+
+// backFiller is the backfilling struct
+type backFiller struct {
+	fetcher       fetcher.StateDiffFetcher
+	batchSize     uint64
+	startingBlock uint64
+}
+
+// NewStorageBackFiller returns a BackFiller
+func NewStorageBackFiller(fetcher fetcher.StateDiffFetcher, startingBlock, batchSize uint64) BackFiller {
+	if batchSize == 0 {
+		batchSize = DefaultMaxBatchSize
+	}
+	return &backFiller{
+		fetcher:       fetcher,
+		batchSize:     batchSize,
+		startingBlock: startingBlock,
+	}
+}
+
+// BackFill fetches, processes, and returns utils.StorageDiffs over a range of blocks
+// It splits a large range up into smaller chunks, batch fetching and processing those chunks concurrently
+func (bf *backFiller) BackFill(endingBlock uint64, backFill chan utils.StorageDiff, errChan chan error, done chan bool) error {
+	if endingBlock < bf.startingBlock {
+		return errors.New("backfill: ending block number needs to be greater than starting block number")
+	}
+	logrus.Infof("going to fill in gap from %d to %d", bf.startingBlock, endingBlock)
+	// break the range up into bins of smaller ranges
+	length := endingBlock - bf.startingBlock + 1
+	numberOfBins := length / bf.batchSize
+	remainder := length % bf.batchSize
+	if remainder != 0 {
+		numberOfBins++
+	}
+	blockRangeBins := make([][]uint64, numberOfBins)
+	for i := range blockRangeBins {
+		nextBinStart := bf.startingBlock + uint64(bf.batchSize)
+		if nextBinStart > endingBlock {
+			nextBinStart = endingBlock + 1
+		}
+		blockRange := make([]uint64, 0, nextBinStart-bf.startingBlock+1)
+		for j := bf.startingBlock; j < nextBinStart; j++ {
+			blockRange = append(blockRange, j)
+		}
+		bf.startingBlock = nextBinStart
+		blockRangeBins[i] = blockRange
+	}
```

Comment:
Sorry to keep hammering away at the same topic but I'm thinking maybe that building bins is another task that could be productively extracted to another function or even namespace. Like what if we had a `BinBuilder` function/struct that took a starting block, ending block, and interval - and then returned the appropriate number of bins, each containing the correct range of blocks? Seems like that could remove a lot of complexity from this namespace and also enable more granular testing of this specific behavior

Reply:
Happy to work on some of this btw if you'd like to split the load - can open up a PR to this branch. Just lmk

Reply:
I'm of the opinion that a seperate namespace and struct for one function is a bit excessive, but a separate helper function makes a lot of sense to me especially since this code is reused in the super node backfiller. I will split it out into a separate function with it's own unit test but if you want to take a step further feel free!

Reply:
Different namespace or not, would appreciate it separated out because this is really hard to follow :)

Reply:
Understood, will split this up a bit!